### PR TITLE
Add a horizontal auto scaler and Metrics

### LINKED ISSUE #11: Add a Horizontal Auto-Scaler
Add a Horizontal Auto-Scaler so a K8s-Cluster can automatically react to heavy load.
Because without it, it could "block" the K8s-Cluster to schedule more Pods.

### LINKED ISSUE #13: Add Prometheus Metrics
Add a Prometheus Metrics Endpoint.

So those Metrics can be used to monitor _Container Pulls_ and they can be used by the HPA (Horizontal Pod Auto-Scaler) mentioned in #11.

### PR DESCRIPTION
**Pull request by l0nax**
This PR adds a Horizontal Auto-Scaler and Metrics to the `kubenab` Application.

The HPA (Horizontal Pod Auto-Scaler) can be configured to work with native K8s Metrics (_CPU Utilization_, etc.) or with Prometheus Metrics (_Request per Second_; _Request Duration_; etc.)

Closes #11 and Closes #13

## TODO

- [x] Add Prometheus Metrics
- [x] Use the CRD's of the Prometheus-Operator to collect the Metrics
- [x] Add Metrics to Helm Chart
- [x] Document usable Metrics of _HPA_ in the Helm-Chart `values.yaml`
  * [x] Done but only in the `values.yaml` ==> _Needs Works_
- [x] Document Usage of the _HPA_

## Changelog

Here is a _Changelog_, which can be later used when releasing a new version.

- **Helm-Chart:** Add HPA with build-in and custom Metrics
- **Helm-Chart:** Add Prometheus Metrics Scraper
- **Helm-Chart:** Add Application Metrics
- **Raw Deployments:** Update `apiVersion` to support deployment on a 1.16+ K8s-Cluster

### CODE REVIEW

**Comment by rimusz** [comment]
@l0nax thanks for your PRs 👍 
Could you please rebase this PR from master, as I did split `main.go` to two files:
- main.go
- admissions.go

Thanks

**Comment by l0nax** [comment]
@rimusz `master` rebased over `feature/11_add-a-horizontal-auto-scaler`

**Comment by rimusz** [comment]
Please move HPA code out from the master and put to `addmissions.go`, thanks

**Comment by l0nax** [comment]
Done

**Comment by l0nax** [comment]
But then we would have Problems with Deployments on 1.16+ K8s-Clusters.
Because the APIs has been marked as deprecated.

https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.16.md#deprecations-and-removals

**Comment by l0nax** [comment]
We could add a check for the Cluster Version and then change the `apiVersion` dynamically.

**Comment by rimusz** [comment]
yes, the ApiVersion needs to be changed dynamically as e.g. cloud providers there is no k8s 1.16 support yet and it is very rear to see k8s 1.16 in prod yet

**Comment by l0nax** [comment]
Added Support for K8s-Clusters with version _<=1.12_ and _>=1.16_

**Comment by l0nax** [comment]
The custom Metric for `Request per Second` in combination with the HPA is implemented and works but the `Request Duration/ Latency` seems to be difficult to implement with the `prometheus-adapter` I think I will need some more time.

Sorry :/

**Comment by rimusz** [comment]
take your time @l0nax and thanks for your contributions 👍

**Comment by l0nax** [comment]
The K8s-Prometheus-Adapter Config now finally works:

ConfigMap entries:
```yaml
  config.yaml: |
    rules:
    - seriesQuery: '{namespace!="",pod!=""}'
      name:
        matches: "^(.*)_milliseconds_bucket"
        as: "http_request_duration"
      resources:
        overrides:
          namespace: {resource: "namespace"}
          pod: {resource: "pod"}
      metricsQuery: 'histogram_quantile(0.99, sum(rate(<<.Series>>[5m])) by (le, api_method, endpoint, <<.GroupBy>>))'
    - seriesQuery: 'http_requests_total{namespace!="",pod!=""}'
      resources:
        overrides:
          namespace: {resource: "namespace"}
          pod: {resource: "pod"}
      name:
        matches: "^(.*)_total"
        as: "${1}_per_second"
      metricsQuery: 'sum(rate(<<.Series>>{<<.LabelMatchers>>}[2m])) by (<<.GroupBy>>)'
```


### Raw API Check

_NOTE: To get "realistic" results in the `custom.metrics` API I run the `benchmark.sh` Script in the background to generate load_

#### Available Metrics

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1" | jq`

```json
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "custom.metrics.k8s.io/v1beta1",
  "resources": [
    {
      "name": "pods/http_request_duration",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "pods/http_requests_per_second",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "namespaces/http_requests_per_second",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "namespaces/http_request_duration",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    }
  ]
}
```

#### Requests per second

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/*/http_requests_per_second" | jq`

```json
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/%2A/http_requests_per_second"
  },
  "items": [
    {
      "describedObject": {
        "kind": "Pod",
        "namespace": "monitoring",
        "name": "kubenab-deployment-6468d895f5-kdvkl",
        "apiVersion": "/v1"
      },
      "metricName": "http_requests_per_second",
      "timestamp": "2019-10-09T19:17:07Z",
      "value": "340767m"
    }
  ]
}
```

#### Response latency

* Command: `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/*/http_request_duration" | jq`

```json
{
  "kind": "MetricValueList",
  "apiVersion": "custom.metrics.k8s.io/v1beta1",
  "metadata": {
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/namespaces/monitoring/pods/%2A/http_request_duration"
  },
  "items": [
    {
      "describedObject": {
        "kind": "Pod",
        "namespace": "monitoring",
        "name": "kubenab-deployment-6468d895f5-kdvkl",
        "apiVersion": "/v1"
      },
      "metricName": "http_request_duration",
      "timestamp": "2019-10-09T19:16:56Z",
      "value": "4952m"
    }
  ]
}
```

**Comment by l0nax** [comment]
The _HPA_ Feature works now perfectlly.

The _HPA_ reacted after 10 seconds.

```bash
~] k describe hpa release-name-kubenab
Name:                                  release-name-kubenab
Namespace:                             monitoring
Labels:                                app.kubernetes.io/instance=release-name
                                       app.kubernetes.io/managed-by=Tiller
                                       app.kubernetes.io/name=kubenab
                                       app.kubernetes.io/version=0.3.2
                                       helm.sh/chart=kubenab-0.0.7
Annotations:                           kubectl.kubernetes.io/last-applied-configuration:
                                         {"apiVersion":"autoscaling/v2beta2","kind":"HorizontalPodAutoscaler","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"...
CreationTimestamp:                     Wed, 09 Oct 2019 21:48:42 +0200
Reference:                             Deployment/kubenab-deployment
Metrics:                               ( current / target )
  "http_requests_per_second" on pods:  568870m / 300
  "http_request_duration" on pods:     4952m / 5
Min replicas:                          1
Max replicas:                          10
Deployment pods:                       1 current / 2 desired
Conditions:
  Type            Status  Reason              Message
  ----            ------  ------              -------
  AbleToScale     True    SucceededRescale    the HPA controller was able to update the target scale to 2
  ScalingActive   True    ValidMetricFound    the HPA was able to successfully calculate a replica count from pods metric http_requests_per_second
  ScalingLimited  False   DesiredWithinRange  the desired count is within the acceptable range
Events:
  Type    Reason             Age   From                       Message
  ----    ------             ----  ----                       -------
  Normal  SuccessfulRescale  15s   horizontal-pod-autoscaler  New size: 2; reason: pods metric http_requests_per_second above target

```

The `deployment` get's scaled up:
```
[...]
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Progressing    True    NewReplicaSetAvailable
  Available      True    MinimumReplicasAvailable
OldReplicaSets:  <none>
NewReplicaSet:   kubenab-deployment-6468d895f5 (2/2 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  97s   deployment-controller  Scaled up replica set kubenab-deployment-6468d895f5 to 2
```

**Comment by l0nax** [comment]
@rimusz The PR is now ready to be reviewed.

File: cmd/kubenab/admissions.go [hunk 1]
```diff
@@ -32,6 +50,14 @@ type patch struct {
 }
 
 func mutateAdmissionReviewHandler(w http.ResponseWriter, r *http.Request) {
+	httpRequestsTotal.With(prometheus.Labels{"api_endpoint": "metrics"}).Inc()
```
Comment (elioengcomp) [inline cmd/kubenab/admissions.go]: Shouldn't the `api_endpoint` here be set to `mutate` instead of `metrics`?

Comment (l0nax) [reply cmd/kubenab/admissions.go]: Oh :sweat_smile:

**Review by elioengcomp — APPROVED** [review]

**Review by rimusz — APPROVED** [review]
LGTM

**Pull request merged**